src/main/scala/viper/silver/plugin/standard/inline/InlineErrorChecker.scala

```diff
@@ -24,6 +26,40 @@ trait InlineErrorChecker {
     recursivePreds
   }
 
+  /**
+    * Construct a call-graph of predicates specified by the given predicate ids. If any predicates
+    * are found to be mutually-recursive, print a warning to the console that it shall not be inlined.
+    * Return a set of predicates that are mutually-recursive.
+    *
+    * @param predicateIds the ids of the predicates we want to inline.
+    * @param program the program for which we are performing predicate inlining on.
+    * @return the set of mutually-recursive predicates.
+    */
+  def checkMutualRecursive(predicateIds: Set[String], program: Program): Set[Predicate] = {
+    val predicatesToInspect = predicateIds.map(program.findPredicate)
+    val predicateCallGraph = {
+      val graph = new DefaultDirectedGraph[Predicate, DefaultEdge](classOf[DefaultEdge])
+      predicatesToInspect.foreach(graph.addVertex)
+
+      def process(pred: Predicate, node: Node): Unit =
+        node.visit {
+          case PredicateAccessPredicate(PredicateAccess(_, name), _) =>
+            graph.addEdge(pred, program.findPredicate(name))
+        }
+
+      predicatesToInspect.foreach { pred =>
+        pred.body.foreach(process(pred, _))
+      }
+      graph
```

Comment:
I don't like that this code is here, feels like it belongs in its own special trait. That said, it's a one-off and I'm not sure if refactoring it is going to make it any better or more correct.